src/components/ChallengeEditor/TimelineTemplate-Field/index.js

```diff
@@ -0,0 +1,120 @@
+import _ from 'lodash'
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import Select from '../../Select'
+import cn from 'classnames'
+import styles from './TimelineTemplate-Field.module.scss'
+
+class TimelineTemplateField extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      validOptions: [],
+      selectedOption: {}
+    }
+
+    this.checkData = this.checkData.bind(this)
+    this.loadSelectedOption = this.loadSelectedOption.bind(this)
+    this.getErrorMessage = this.getErrorMessage.bind(this)
+  }
+
+  componentDidMount () {
+    this.checkData()
+  }
+
+  shouldComponentUpdate (nextProps) {
+    const { onUpdateSelect, challengeTimelines, timelineTemplates, challenge } = nextProps
+    const hasSelectedTypeAndTrack = !_.isEmpty(challenge.typeId) && !_.isEmpty(challenge.trackId)
+    if ((hasSelectedTypeAndTrack && this.state.validOptions.length === 0) || this.state.matchString !== `${challenge.typeId}-${challenge.trackId}-${this.state.selectedOption.value}`) {
+      this.checkData(onUpdateSelect, challengeTimelines, timelineTemplates, challenge)
+    }
+    return true
+  }
```

Comment:
It seems that `shouldComponentUpdate` is used as a replacement for *unsafe*`componentWillUpdate` and it triggers `this.setState` and external callback `onUpdateSelect` inside. As per my understanding such approach is not recommended and in particular it could lead to a circular re-reredering.

Here is some guide on which approaches are recommended as a replacement for unsafe methods https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html.  
It includes examples for [updating state based on props](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) and [invoking external callbacks](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#invoking-external-callbacks).